Interop/Runtime/src/jvm/kotlin/kotlinx/cinterop/JvmNativeMem.kt

```diff
@@ -95,17 +95,79 @@ internal object nativeMemUtils {
         return unsafe.allocateInstance(T::class.java) as T
     }
 
+    private fun alignUp(x: Long, align: Int) = (x + align - 1) and (align - 1).toLong().inv()
+    private fun alignUp(x: Int, align: Int) = (x + align - 1) and (align - 1).inv()
+
+    // 256 buckets for sizes <= 2048 padded to 8
+    // 256 buckets for sizes <= 64KB padded to 256
+    // 256 buckets for sizes <= 1MB padded to 4096
+    private val smallChunks = LongArray(256)
+    private val mediumChunks = LongArray(256)
+    private val bigChunks = LongArray(256)
+
+    // Chunk layout: [chunk size,...padding...,diff to start,aligned data start,.....data.....]
     fun alloc(size: Long, align: Int): NativePointed {
-        val address = unsafe.allocateMemory(
-                if (size == 0L) 1L else size // It is a hack: `sun.misc.Unsafe` can't allocate zero bytes
-        )
+        val chunkHeaderSize = 8 // chunk size + alignment hop size.
+        val totalChunkSize = chunkHeaderSize + size + align
+        val ptr = chunkHeaderSize + when {
+            totalChunkSize <= 2048 -> allocFromFreeList(totalChunkSize.toInt(), 8, smallChunks)
```

Comment:
May be save save all these numbers as named constant fields in case if they are changed in future? They all are repeated both in allocation and free functions.
```
private const val chunkHeaderSize = 8 (may be even 2 * Int.SIZE_BYTES)
private const val maxSmallChunkSize = 2048
...
```

May be also replacing 4 with `Int.SIZE_BYTES` made this code a bit easier for me at least.

Reply:
Also may be better to connect padding with long array sizes `paddingSize = maxSmallChunkSize / arraySize`